### PR TITLE
Fix: Return error and avoid panicking in SetOpener function

### LINKED ISSUE #32989: Panic in html/browsers/the-window-object/window-opener-unconfigurable.window.js
To reproduce: `./mach test-wpt tests/wpt/tests/html/browsers/the-window-object/window-opener-unconfigurable.window.js`

Backtrace:
```
pid:50037 assertion failed: JS_DefineProperty(*cx, obj, c"opener".as_ptr(), value,
 0:02.90 pid:50037     JSPROP_ENUMERATE as u32) (thread Script(1,1), at components/script/dom/window.rs:732)
 0:05.15 pid:50037   13: <script::dom::window::Window as script::dom::bindings::codegen::Bindings::WindowBinding::Window_Binding::WindowMethods>::SetOpener
 0:05.16 pid:50037              at /Users/jdm/src/servo/components/script/dom/window.rs:732:13
 0:05.19 pid:50037   14: script::dom::bindings::codegen::Bindings::WindowBinding::Window_Binding::set_opener::{{closure}}
 0:05.19 pid:50037              at /Users/jdm/src/servo/target/debug/build/script-691217678d484feb/out/Bindings/WindowBinding.rs:1282:26
 0:05.19 pid:50037   15: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
 0:05.19 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:305:13
 0:05.20 pid:50037   16: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:05.20 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/panic/unwind_safe.rs:272:9
 0:05.20 pid:50037   17: std::panicking::try::do_call
 0:05.20 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:559:40
 0:05.20 pid:50037   18: ___rust_try
 0:05.20 pid:50037   19: std::panicking::try
 0:05.20 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:523:19
 0:05.20 pid:50037   20: std::panic::catch_unwind
 0:05.20 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panic.rs:149:14
 0:05.20 pid:50037   21: mozjs::panic::wrap_panic
 0:05.20 pid:50037              at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/c715ae4/mozjs/src/panic.rs:22:11
 0:05.23 pid:50037   22: script::dom::bindings::codegen::Bindings::WindowBinding::Window_Binding::set_opener
 0:05.23 pid:50037              at /Users/jdm/src/servo/target/debug/build/script-691217678d484feb/out/Bindings/WindowBinding.rs:1278:5
 0:05.28 pid:50037   23: CallJitSetterOp
 0:05.28 pid:50037              at /Users/runner/work/mozjs/mozjs/mozjs-sys/src/jsglue.cpp:631:10
 0:05.31 pid:50037   24: script::dom::bindings::utils::call_setter
 0:05.31 pid:50037              at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:567:9
 0:05.34 pid:50037   25: script::dom::bindings::utils::generic_call
 0:05.34 pid:50037              at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:522:5
 0:05.34 pid:50037   26: script::dom::bindings::utils::generic_setter
 0:05.34 pid:50037              at /Users/jdm/src/servo/components/script/dom/bindings/utils.rs:580:5
 0:05.34 pid:50037   27: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.34 pid:50037   28: __ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:05.34 pid:50037   29: __ZN2js10CallSetterEP9JSContextN2JS6HandleINS2_5ValueEEES5_S5_
 0:05.34 pid:50037   30: __ZL19SetExistingPropertyP9JSContextN2JS6HandleINS1_11PropertyKeyEEENS2_INS1_5ValueEEES6_NS2_IPN2js12NativeObjectEEERKNS7_14PropertyResultERNS1_14ObjectOpResultE
 0:05.34 pid:50037   31: __ZN2js17NativeSetPropertyILNS_13QualifiedBoolE1EEEbP9JSContextN2JS6HandleIPNS_12NativeObjectEEENS5_INS4_11PropertyKeyEEENS5_INS4_5ValueEEESC_RNS4_14ObjectOpResultE
 0:05.37 pid:50037   32: script::dom::windowproxy::set
 0:05.37 pid:50037              at /Users/jdm/src/servo/components/script/dom/windowproxy.rs:1016:5
 0:05.37 pid:50037   33: _ZNK19WrapperProxyHandler3setEP9JSContextN2JS6HandleIP8JSObjectEENS3_INS2_11PropertyKeyEEENS3_INS2_5ValueEEESA_RNS2_14ObjectOpResultE
 0:05.37 pid:50037              at /Users/runner/work/mozjs/mozjs/mozjs-sys/src/jsglue.cpp:398:3
 0:05.37 pid:50037   34: __ZN2js5Proxy3setEP9JSContextN2JS6HandleIP8JSObjectEENS4_INS3_11PropertyKeyEEENS4_INS3_5ValueEEESB_RNS3_14ObjectOpResultE
 0:05.37 pid:50037   35: __ZN2js9InterpretEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   36: __ZN2js9RunScriptEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   37: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   38: __ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:05.37 pid:50037   39: __ZN2js8fun_callEP9JSContextjPN2JS5ValueE
 0:05.37 pid:50037   40: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   41: __ZN2js9InterpretEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   42: __ZN2js9RunScriptEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   43: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   44: __ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:05.37 pid:50037   45: __ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
 0:05.37 pid:50037   46: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   47: __ZN2js9InterpretEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   48: __ZN2js9RunScriptEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   49: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   50: __ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:05.37 pid:50037   51: __ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
 0:05.37 pid:50037   52: __ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:05.37 pid:50037   53: __ZN2js9InterpretEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   54: __ZN2js9RunScriptEP9JSContextRNS_8RunStateE
 0:05.37 pid:50037   55: __ZN2js7ExecuteEP9JSContextN2JS6HandleIP8JSScriptEENS3_IP8JSObjectEENS2_13MutableHandleINS2_5ValueEEE
 0:05.37 pid:50037   56: __Z16JS_ExecuteScriptP9JSContextN2JS6HandleIP8JSScriptEENS1_13MutableHandleINS1_5ValueEEE
 0:05.40 pid:50037   57: mozjs::rust::wrappers::JS_ExecuteScript
 0:05.40 pid:50037              at /Users/jdm/.cargo/git/checkouts/mozjs-fa11ffc7d4f1cc2d/c715ae4/mozjs/src/rust.rs:1070:51
 0:05.43 pid:50037   58: script::dom::globalscope::GlobalScope::evaluate_script_on_global_with_result::{{closure}}
 0:05.43 pid:50037              at /Users/jdm/src/servo/components/script/dom/globalscope.rs:2723:34
 0:05.46 pid:50037   59: profile_traits::time::profile
 0:05.46 pid:50037              at /Users/jdm/src/servo/components/shared/profile/time.rs:147:15
 0:05.49 pid:50037   60: script::dom::globalscope::GlobalScope::evaluate_script_on_global_with_result
 0:05.49 pid:50037              at /Users/jdm/src/servo/components/script/dom/globalscope.rs:2652:9
 0:05.52 pid:50037   61: script::dom::htmlscriptelement::HTMLScriptElement::run_a_classic_script
 0:05.52 pid:50037              at /Users/jdm/src/servo/components/script/dom/htmlscriptelement.rs:1057:9
 0:05.52 pid:50037   62: script::dom::htmlscriptelement::HTMLScriptElement::execute
 0:05.52 pid:50037              at /Users/jdm/src/servo/components/script/dom/htmlscriptelement.rs:1019:17
 0:05.54 pid:50037   63: script::dom::servoparser::ServoParser::resume_with_pending_parsing_blocking_script
 0:05.54 pid:50037              at /Users/jdm/src/servo/components/script/dom/servoparser/mod.rs:328:9
 0:05.57 pid:50037   64: script::dom::document::Document::process_pending_parsing_blocking_script
 0:05.57 pid:50037              at /Users/jdm/src/servo/components/script/dom/document.rs:2511:13
 0:05.57 pid:50037   65: script::dom::document::Document::pending_parsing_blocking_script_loaded
 0:05.57 pid:50037              at /Users/jdm/src/servo/components/script/dom/document.rs:2497:9
 0:05.57 pid:50037   66: script::dom::htmlscriptelement::finish_fetching_a_classic_script
 0:05.57 pid:50037              at /Users/jdm/src/servo/components/script/dom/htmlscriptelement.rs:318:13
 0:05.57 pid:50037   67: <script::dom::htmlscriptelement::ClassicContext as net_traits::FetchResponseListener>::process_response_eof
 0:05.57 pid:50037              at /Users/jdm/src/servo/components/script/dom/htmlscriptelement.rs:460:9
 0:05.61 pid:50037   68: <net_traits::FetchResponseMsg as net_traits::Action<T>>::process
 0:05.61 pid:50037              at /Users/jdm/src/servo/components/shared/net/lib.rs:304:25
 0:05.64 pid:50037   69: <script::network_listener::ListenerTask<A,Listener> as script::task::TaskOnce>::run_once
 0:05.64 pid:50037              at /Users/jdm/src/servo/components/script/network_listener.rs:117:13
 0:05.67 pid:50037   70: <script::task::CancellableTask<T> as script::task::TaskOnce>::run_once
 0:05.67 pid:50037              at /Users/jdm/src/servo/components/script/task.rs:114:13
 0:05.67 pid:50037   71: <T as script::task::TaskBox>::run_box
 0:05.67 pid:50037              at /Users/jdm/src/servo/components/script/task.rs:57:9
 0:05.69 pid:50037   72: script::script_thread::ScriptThread::handle_msg_from_script
 0:05.69 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:2480:17
 0:05.72 pid:50037   73: script::script_thread::ScriptThread::handle_msgs::{{closure}}
 0:05.72 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:1958:46
 0:05.72 pid:50037   74: script::script_thread::ScriptThread::profile_event
 0:05.72 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:2214:13
 0:05.73 pid:50037   75: script::script_thread::ScriptThread::handle_msgs
 0:05.73 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:1951:26
 0:05.73 pid:50037   76: script::script_thread::ScriptThread::start
 0:05.73 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:1454:15
 0:05.73 pid:50037   77: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}::{{closure}}
 0:05.73 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:845:25
 0:05.75 pid:50037   78: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:05.75 pid:50037              at /Users/jdm/src/servo/components/shared/profile/mem.rs:91:9
 0:05.75 pid:50037   79: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}
 0:05.75 pid:50037              at /Users/jdm/src/servo/components/script/script_thread.rs:843:17
 0:05.78 pid:50037   80: std::sys_common::backtrace::__rust_begin_short_backtrace
 0:05.78 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/sys_common/backtrace.rs:155:18
 0:05.81 pid:50037   81: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
 0:05.81 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/thread/mod.rs:542:17
 0:05.84 pid:50037   82: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:05.84 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/panic/unwind_safe.rs:272:9
 0:05.87 pid:50037   83: std::panicking::try::do_call
 0:05.87 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:559:40
 0:05.87 pid:50037   84: ___rust_try
 0:05.87 pid:50037   85: std::panicking::try
 0:05.87 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:523:19
 0:05.87 pid:50037   86: std::panic::catch_unwind
 0:05.87 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panic.rs:149:14
 0:05.87 pid:50037       std::thread::Builder::spawn_unchecked_::{{closure}}
 0:05.87 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/thread/mod.rs:541:30
 0:06.23 pid:50037   87: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:06.23 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:250:5
 0:06.23 pid:50037   88: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:06.23 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/alloc/src/boxed.rs:2063:9
 0:06.23 pid:50037       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:06.23 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/alloc/src/boxed.rs:2063:9
 0:06.23 pid:50037       std::sys::pal::unix::thread::Thread::new::thread_start
 0:06.23 pid:50037              at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/sys/pal/unix/thread.rs:108:17
 0:06.23 pid:50037   89: __pthread_joiner_wake
```

### PR DESCRIPTION
**Pull request by Taym95**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #32989

- [x] These changes do not require tests because there is not behavior change.

### CODE REVIEW

File: components/script/dom/window.rs [hunk 1]
```diff
@@ -702,41 +702,43 @@ impl WindowMethods for Window {
     }
 
     // https://html.spec.whatwg.org/multipage/#dom-opener
-    fn Opener(&self, cx: JSContext, in_realm_proof: InRealm) -> JSVal {
```
Comment (jdm) [inline components/script/dom/window.rs:705]: Huh, the missing InRealm argument looks like a codegen bug to me. I'll file a followup issue about it.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
Looks like the test is still panicking, but with a different exception. Did you try running it locally?
```
58229:{"action": "process_output", "time": 1723408618902, "thread": "ProcessReader", "pid": 30829, "source": "TestRunner-0", "process": "30925", "data": "assertion failed: !JS_IsExceptionPending(*cx) (thread Script(1,1), at components/script/dom/bindings/error.rs:160)", "command": "/home/runner/work/servo/servo/target/release/servo --hard-fail -u Servo/wptrunner --ignore-certificate-errors -z http://web-platform.test:8000/html/browsers/the-window-object/window-opener-unconfigurable.window.html -Z replace-surrogates --user-stylesheet /home/runner/work/servo/servo/resources/ahem.css --certificate-path /home/runner/work/servo/servo/tests/wpt/tests/tools/certs/cacert.pem", "test": "/html/browsers/the-window-object/window-opener-unconfigurable.window.html"}
```
I suspect you need to throw the JsFailed exception, rather than InvalidState.

**Comment by Taym95** [comment]
> Looks like the test is still panicking, but with a different error message. Did you try running it locally?
> 
> ```
> 58229:{"action": "process_output", "time": 1723408618902, "thread": "ProcessReader", "pid": 30829, "source": "TestRunner-0", "process": "30925", "data": "assertion failed: !JS_IsExceptionPending(*cx) (thread Script(1,1), at components/script/dom/bindings/error.rs:160)", "command": "/home/runner/work/servo/servo/target/release/servo --hard-fail -u Servo/wptrunner --ignore-certificate-errors -z http://web-platform.test:8000/html/browsers/the-window-object/window-opener-unconfigurable.window.html -Z replace-surrogates --user-stylesheet /home/runner/work/servo/servo/resources/ahem.css --certificate-path /home/runner/work/servo/servo/tests/wpt/tests/tools/certs/cacert.pem", "test": "/html/browsers/the-window-object/window-opener-unconfigurable.window.html"}
> ```
> 
> I suspect you need to throw the JsFailed exception, rather than InvalidState.

Should be passing now, thanks for review.

File: components/script/dom/window.rs (hunk 1 above)
Comment (Taym95) [reply components/script/dom/window.rs:705]: indeed it is weird, I just added `Throws` and `in_realm_proof ` was removed from codegen!

**Review by jdm — APPROVED** [review]

**Review by gterzian — CHANGES_REQUESTED** [review]

**Review by gterzian — APPROVED** [review]

**Review by gterzian — CHANGES_REQUESTED** [review]
We should file an issue for a follow-up for https://github.com/servo/servo/pull/33002/files#r1713297324

**Review by gterzian — APPROVED** [review]
We should file an issue for a follow-up for https://github.com/servo/servo/pull/33002/files#r1713297324

**Comment by Taym95** [comment]
> We should file an issue for a follow-up for https://github.com/servo/servo/pull/33002/files#r1713297324

I create the issue, I will try to follow up.

**Pull request merged**